src/grouping_map.rs

```diff
@@ -0,0 +1,562 @@
+
+/// `GroupingMap` is an intermediate struct for efficient group-and-fold operations.
+/// It groups elements by their key and at the same time fold each group
+/// using some aggregating operation.
+/// 
+/// No method on this struct performs temporary allocations.
+#[must_use = "GroupingMap is lazy and do nothing unless consumed"]
+pub struct GroupingMap<I> {
+    iter: I,
+}
+
+impl<I, K, V> GroupingMap<I>
+    where I: Iterator<Item = (K, V)>,
+          K: Hash + Eq,
+{
+    /// This is the generic way to perform any operation on a `GroupingMap`.
+    /// It's suggested to use this method only to implement custom operations
+    /// when the already provided ones are not enough.
+    /// 
+    /// Groups elements from the `GroupingMap` source by key and applies `operation` to the elements
+    /// of each group sequentially, passing the previously accumulated value, a reference to the key
+    /// and the current element as arguments, and stores the results in an `HashMap`.
+    ///
+    /// The `operation` function is invoked on each element with the following parameters:
+    ///  - the current value of the accumulator of the group if there is currently one;
+    ///  - a reference to the key of the group this element belongs to;
+    ///  - the element from the source being aggregated;
+    /// 
+    /// If `operation` returns `Some(element)` then the accumulator is updated with `element`,
+    /// otherwise the previous accumulation is discarded.
+    ///
+    /// Return a `HashMap` associating the key of each group with the result of aggregation of
+    /// that group's elements. If the aggregation of the last element of a group discards the
+    /// accumulator then there won't be an entry associated to that group's key.
+    /// 
+    /// ```
+    /// use itertools::Itertools;
+    /// 
+    /// let data = vec![2, 8, 5, 7, 9, 0, 4, 10];
+    /// let lookup = data.into_iter()
+    ///     .into_grouping_map_by(|&n| n % 4)
+    ///     .aggregate(|acc, _key, val| {
+    ///         if val == 0 || val == 10 {
+    ///             None
+    ///         } else {
+    ///             Some(acc.unwrap_or(0) + val)
+    ///         }
+    ///     });
+    /// 
+    /// assert_eq!(lookup[&0], 4);        // 0 resets the accumulator so only 4 is summed
+    /// assert_eq!(lookup[&1], 5 + 9);
+    /// assert_eq!(lookup.get(&2), None); // 10 resets the accumulator and nothing is summed afterward
+    /// assert_eq!(lookup[&3], 7);
+    /// assert_eq!(lookup.len(), 3);      // The final keys are only 0, 1 and 2
+    /// ```
+    pub fn aggregate<FO, R>(self, mut operation: FO) -> HashMap<K, R>
+        where FO: FnMut(Option<R>, &K, V) -> Option<R>,
+    {
+        let mut destination_map = HashMap::new();
+
+        for (key, val) in self.iter {
+            let acc = destination_map.remove(&key);
+            if let Some(op_res) = operation(acc, &key, val) {
+                destination_map.insert(key, op_res);
+            }
```

Comment:
Using `FnMut(Option<R>, ...)` forces the caller to deal with `Option`, i.e. the caller has to deal with `Some`/`None`, essentially checking the same thing that is already checked in `aggregate` in terms of existence in `destination_map`. It may be ok, but we could avoid these checks in the callers by having two functions (one for the `None`-case, one for the `Some`-case).

Reply:
> Do we really need to `remove` and `insert` back into `destination_map`? Couldn't we facilitate the `entry`-API?

`remove` and `insert` are needed if we want to pass an owned accumulator to the closure, like the normal `Iterator::fold`. The alternative is passing a mutable reference but this is more restricting for the end user.
Maybe I should add a second `aggregate` function that provides mutable access instead? This would allow a more efficient implementation in some cases, like in `GroupingMap::collect` (currently it duplicates `aggregate`'s code).

> Using `FnMut(Option<R>, ...)` forces the caller to deal with `Option`, i.e. the caller has to deal with `Some`/`None`, essentially checking the same thing that is already checked in `aggregate` in terms of existence in `destination_map`. It may be ok, but we could avoid these checks in the callers by having two functions (one for the `None`-case, one for the `Some`-case).

Wouldn't this just move the check from the closure to the `aggregate` function?
Performance wise it should remain the same.
But what if the caller wants to mutate an external variable in both the case the map contains the value or not? It can't do that with 2 closures.

Reply:
I was thinking about this, in particular since issue #467. I made a couple of basic benchmarks and the diffeference in speed is not negligible. For example on a simple `collect` of 1_000_000 `i32` grouped by their reminder by 10 or 10_000, the entry version performed 2x better! But at what cost? 

- I had to add another closure for the init.
- The `init` closure can't take a reference to the `key` as parameter because `Entry::or_insert_with_key` is unstable. Not sure but there may be a way to do it anyway.
- I had to remove the `key` parameter from the other closure as well
- I had to pass a mutable reference to the accumulator instead of an owned value because the `Entry` api gives me only that.

The problem is that the current approach move a lot the accumulator and makes two lookup in the map, one for `remove` and one for `insert`. Moving the accumulator is required if we don't want to restrict the API, so I wondered if I could use the `OccupiedEntry` and `VacantEntry` API to make only one lookup, but unfortunately currently it's impossible. If `OccupiedEntry` had a way to remove the entry, returning the value, but also returning the just freed `VacantEntry` it would be perfect, but there's no such API. I tried to imitate it using `unsafe` and `MaybeUninit` and this is the result: 

https://gist.github.com/SkiFire13/12a61302e35ffe52c0337e0959fb15f6

On the previous example it runs 1.6x faster than the current approach but 1.3 slower than the normal `entry`. On another example (count the number of occurences of a bunch of `i32`s) it performed 2.5x the current approach and the same as the normal entry. This could actually be used to unify the implementation with #468, however I really don't like that I had to use `unsafe` for this. There's also the problem that `MaybeUninit` wasn't stable in rust 1.32.0

Or maybe should I just add an `aggregate_mut`/`aggregate_update` (has anyone better names?) that provides a restricted but faster API using `entry`?